src/algorithm/bisection_loop!.jl

```diff
@@ -0,0 +1,335 @@
+                sys,
+                res_hi,
+                BaseMVA,
+                lmp_hi,
+                tcrdd_slack_hi,
+                bid_hi,
+                Pmin_orig,
+                Pmax_orig;
+                segm_bid_argmax_profit,
+                print_plots
+            )
+
+            # Check if the inverse tangent of the RDC is the same for both functions
+            comparison_res = compare_tan_inv_RDC(
+                sys,
+                BaseMVA,
+                res_lo,
+                res_hi,
+                lmp_lo,
+                lmp_hi,
+                tcrdd_slack_lo,
+                tcrdd_slack_hi,
+                bid_lo,
+                bid_hi,
+                Pmin_orig,
+                Pmax_orig;
+                segm_bid_argmax_profit,
+                epsilon,
+                print_plots
+            )
+            # Unpack the results of the comparison
+            same_tan_inv_lo_hi, intersect_tan_inv_lo_hi, bid_intersect = comparison_res
+
+            # Find bid_mid depending on the the case
+            if same_tan_inv_lo_hi
+                #tan_inv_RDC is the same for both hi and lo
+                bid_mid = bid_argmax_lo
+            elseif !intersect_tan_inv_lo_hi ||
+                    ( intersect_tan_inv_lo_hi &&
+                        (bid_intersect < bid_lo || bid_hi < bid_intersect) )
+                # They dont intersect or intersect out of range, do traditional bisection
+                bid_mid = 0.5*(bid_hi + bid_lo)
+            else
+                # Functions are different but they intersect within range
+                # Determine profit maximiser range (Table 1 of reference [1])
+                # Check where is the hump
+                belong_bid_lo_intsct = false
+                belong_bid_hi_intsct = false
+                if (bid_lo <= bid_argmax_lo && bid_argmax_lo <= bid_intersect)
+                    belong_bid_lo_intsct = true
+                end
+                if (bid_intersect <= bid_argmax_hi && bid_argmax_hi <= bid_hi)
+                    belong_bid_hi_intsct = true
+                end
+                # Select bid_mid
+                if !belong_bid_lo_intsct && !belong_bid_hi_intsct #no hump
+                    bid_mid = bid_intersect
+                elseif belong_bid_lo_intsct && !belong_bid_hi_intsct #left hump
+                    bid_mid = bid_argmax_lo
+                elseif !belong_bid_lo_intsct && belong_bid_hi_intsct #right hump
+                    bid_mid = bid_argmax_hi
+                elseif belong_bid_lo_intsct && belong_bid_hi_intsct #double hump
+                    # Select the side based on profit_argmax
+                    if profit_argmax_hi >= profit_argmax_lo
+                        bid_mid = bid_argmax_hi
+                    else
+                        bid_mid = bid_argmax_lo
+                    end
+                end
+            end
+        end
```

Comment:
I feel like large blocks like this can be broken down into their own functions. It'll allow for a few things:

- Simplify reading large functions and maintaining them
- Provides space for providing more comments (what are we trying to set here with all these `if` statements)
- Easier testing as you don't need to run through all the code before/after